metron-platform/metron-writer/src/main/java/org/apache/metron/writer/BulkWriterComponent.java

```diff
@@ -326,7 +328,6 @@ public void flushTimeouts(
           || clock.currentTimeMillis() - batchTimeoutInfo[LAST_CREATE_TIME_MS] >= batchTimeoutInfo[TIMEOUT_MS]) {
         flush(sensorType, bulkMessageWriter, configurations, messageGetStrategy
```

Comment:
flush will remove an entry from the 'sensorTupleMap' and we are iterating on the same collection using iterator, the iterator.next() will throw ConcurrentModificationException

Reply:
Ouch.  Good catch.  This may be a long-standing race condition hidden by the  faulty timeout logic.

Reply:
Rescinding my +1...  wa wa wahh.

Reply:
here is the stack trace for the same , 

Caused by: java.util.ConcurrentModificationException
	at java.util.HashMap$HashIterator.nextNode(HashMap.java:1437) ~[?:1.8.0_112]
	at java.util.HashMap$KeyIterator.next(HashMap.java:1461) ~[?:1.8.0_112]
	at org.apache.metron.writer.BulkWriterComponent.flushTimeouts(BulkWriterComponent.java:278) ~[stormjar.jar:?]
	at org.apache.metron.writer.bolt.BulkMessageWriterBolt.execute(BulkMessageWriterBolt.java:221) ~[stormjar.jar:?]
	at org.apache.storm.daemon.executor$fn__10195$tuple_action_fn__10197.invoke(executor.clj:735) ~[storm-core-1.1.0.2.6.5.0-292.jar:1.1.0.2.6.5.0-292]
	at org.apache.storm.daemon.executor$mk_task_receiver$fn__10114.invoke(executor.clj:469) ~[storm-core-1.1.0.2.6.5.0-292.jar:1.1.0.2.6.5.0-292]
	at org.apache.storm.disruptor$clojure_handler$reify__4137.onEvent(disruptor.clj:40) ~[storm-core-1.1.0.2.6.5.0-292.jar:1.1.0.2.6.5.0-292]
	at org.apache.storm.utils.DisruptorQueue.consumeBatchToCursor(DisruptorQueue.java:472) ~[storm-core-1.1.0.2.6.5.0-292.jar:1.1.0.2.6.5.0-292]
	... 6 more

Reply:
Good catch @MohanDV.  The latest commit should address this.

Reply:
Can you add to the comment with info about the ConcurrentModificationException? Without knowing why we're creating a copy of the keyset, it's not obvious.

Reply:
Done.